### PR TITLE
[ECO-5114] Provide mechanism for waiting to be able to perform presence operations

### LINKED ISSUE #112: Provide mechanism for waiting to be able to perform presence operations
That is, to be able to implement the following:

```
** @(CHA-PR3d)@ @[Testable]@ If the room status is @Attaching@, then the @enter@ call will wait for the attach operation to succeed and then call the underlying presence @enter@ operation. It will throw an error if the attach fails.
** @(CHA-PR3e)@ @[Testable]@ If the room status is @Attached@, then the @enter@ call will invoke the underlying @presence.enter()@ call.
** @(CHA-PR3f)@ @[Testable]@ If the room status is @Detached@, an @ErrorInfo@ with code 40000, explaining that attach must be called first, will be thrown.
** @(CHA-PR3g)@ @[Testable]@ For any other room status, an @ErrorInfo@ with code 40000 will be thrown.
```

and the equivalent in:

- CHA-PR10d - g
- CHA-PR6c - f
- CHA-T2c - f

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-5114) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
**Note: This PR is based on top of #118; please review that one first.**

This should unblock @umair-ably’s implementation of presence and typing indicators. See commit messages for more details.

Resolves #112.

## Summary by CodeRabbit

- **New Features**
	- Introduced new error handling cases for better clarity in error messages.
	- Added methods to manage presence operations based on room status.
	- Enhanced lifecycle management for feature channels.
	- Updated `Helpers` function to validate error messages.
	- Expanded `MockFeatureChannel` to handle presence operation readiness.

- **Bug Fixes**
	- Improved error handling and state management in the Room Lifecycle Manager.

- **Tests**
	- Added new test cases for presence operation behaviors and lifecycle management.
	- Updated existing tests to reflect changes in method signatures and behaviors.

### CODE REVIEW

File: Tests/AblyChatTests/Mocks/MockFeatureChannel.swift
Comment (lawrence-forooghian) [reply Tests/AblyChatTests/Mocks/MockFeatureChannel.swift:30]: Please learn about Swift’s typed throws syntax

File: Tests/AblyChatTests/Mocks/MockFeatureChannel.swift
Comment (lawrence-forooghian) [reply Tests/AblyChatTests/Mocks/MockFeatureChannel.swift:15]: Will fix

File: Tests/AblyChatTests/Mocks/MockRoomLifecycleManager.swift
Comment (lawrence-forooghian) [reply Tests/AblyChatTests/Mocks/MockRoomLifecycleManager.swift:61]: Will implement when needed

File: Sources/AblyChat/RoomLifecycleManager.swift
Comment (maratal) [inline Sources/AblyChat/RoomLifecycleManager.swift:713]: If `waitForCompletionOfOperationWithID` throws nothing will happen?

File: Sources/AblyChat/RoomLifecycleManager.swift
Comment (maratal) [inline Sources/AblyChat/RoomLifecycleManager.swift:912]: Why forced here?

File: Sources/AblyChat/RoomLifecycleManager.swift
Comment (maratal) [inline Sources/AblyChat/RoomLifecycleManager.swift:930]: And here

**Review by maratal — COMMENTED** [review]
Left a few questions

File: Sources/AblyChat/RoomLifecycleManager.swift
Comment (lawrence-forooghian) [reply Sources/AblyChat/RoomLifecycleManager.swift:713]: Correct — CHA-RL1d just says "** @(CHA-RL1d)@ @[Testable]@ If a room lifecycle operation is already in progress, this operation shall wait until the current operation completes before continuing"; it doesn't say to use the result of that operation.

File: Sources/AblyChat/RoomLifecycleManager.swift
Comment (lawrence-forooghian) [reply Sources/AblyChat/RoomLifecycleManager.swift:912]: This is a result of the change in b58f9d783211026201c5c6d5ab1ed1361ea874e8 to make `waitForCompletionOfOperationWithID` have a typed throw (i.e. to declare that it throws `ARTErrorInfo` instead of just `Error`); it meant that I had to change `performAnOperation` to always throw `ARTErrorInfo` instead of only throwing an error if the `body` closure throws an error.

Comment (lawrence-forooghian) [reply Sources/AblyChat/RoomLifecycleManager.swift:912]: (There might be a better way of handling all this stuff; I hope so. But this is the best I could do in the limited time I have now.)

Comment (maratal) [reply Sources/AblyChat/RoomLifecycleManager.swift:912]: I think you should just mark `performReleaseOperation...` funcs as throwing then.

Comment (maratal) [reply Sources/AblyChat/RoomLifecycleManager.swift:912]: But it's a minor, just make sure it will be addressed later (maybe create an issue for resolving all `swiftlint:disable:next force_try`)

**Review by maratal — APPROVED** [review]
LGTM

**Review by umair-ably — APPROVED** [review]
LGTM - happy to review again after resolving merge conflicts

**Pull request merged**
